### PR TITLE
Tensor::Mean

### PR DESCRIPTION
**Pull request by yxlao**
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/1797)

### CODE REVIEW

File: src/UnitTest/Python/test_core.py [hunk 1]
```diff
@@ -536,6 +536,19 @@ def test_reduction_sum(dim, keepdim):
...
+@pytest.mark.parametrize(
+    "dim",
+    [0, 1, 2, (), (0,), (1,), (2,), (0, 1), (0, 2), (1, 2), (0, 1, 2), None])
+@pytest.mark.parametrize("keepdim", [True, False])
+def test_reduction_mean(dim, keepdim):
+    np_src = np.array(range(24)).reshape((2, 3, 4)).astype(np.float32)
+    o3_src = o3d.Tensor(np_src)
+
+    np_dst = np_src.mean(axis=dim, keepdims=keepdim)
+    o3_dst = o3_src.mean(dim=dim, keepdim=keepdim)
```
Comment (pvarvak) [inline src/UnitTest/Python/test_core.py:548]: can we also test o3_src.cuda().mean (if we have cuda?)

File: src/UnitTest/Core/Tensor.cpp [hunk 2]
```diff
@@ -1247,6 +1267,84 @@ TEST_P(TensorPermuteDevices, ReduceSumNotKeepDim) {
...
+    // np.sum(np.ones((0, 2)), axis=(1,), keepdims=*)
+    src = Tensor::Ones({0, 2}, Dtype::Float32, device);
+    dst = src.Sum({1}, false);
+    EXPECT_EQ(dst.GetShape(), SizeVector({0}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({}));
+    dst = src.Sum({1}, true);
+    EXPECT_EQ(dst.GetShape(), SizeVector({0, 1}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({}));
+
+    // np.sum(np.ones((0, 2)), axis=(0, 1), keepdims=*), fill with identity
```
Comment (pvarvak) [inline src/UnitTest/Core/Tensor.cpp:1338]: can we add (some of) these to test_core.py (and compare to numpy)?

File: src/Open3D/Core/Tensor.cpp [hunk 3]
```diff
@@ -681,6 +681,27 @@ Tensor Tensor::Sum(const SizeVector& dims, bool keepdim) const {
...
+Tensor Tensor::Mean(const SizeVector& dims, bool keepdim) const {
+    if (dtype_ != Dtype::Float32 && dtype_ != Dtype::Float64) {
+        utility::LogError(
+                "Can only compute mean for Float32 or Float64, got {} instead.",
+                DtypeUtil::ToString(dtype_));
+    }
+
+    Tensor dst(shape_util::ReductionShape(shape_, dims, keepdim), dtype_,
+               GetDevice());
+    kernel::Reduction(*this, dst, dims, keepdim, kernel::ReductionOpCode::Sum);
```
[the remark this answers is not in the document]
Comment (yxlao) [reply src/Open3D/Core/Tensor.cpp:693]: done

File: src/UnitTest/Core/Tensor.cpp [hunk 4]
```diff
@@ -2138,3 +2236,53 @@ TEST_P(TensorPermuteDevices, ScalarOperatorOverload) {
...
+    src = Tensor(std::vector<float>({0, 1, 2, 3, 4, 5}), {2, 3}, Dtype::Float32,
+                 device);
+    dst = src.Mean({}, false);
+    EXPECT_EQ(dst.GetShape(), SizeVector({2, 3}));
+    EXPECT_EQ(dst.ToFlatVector<float>(),
+              std::vector<float>({0, 1, 2, 3, 4, 5}));
+    dst = src.Mean({0}, false);
+    EXPECT_EQ(dst.GetShape(), SizeVector({3}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({1.5, 2.5, 3.5}));
+    dst = src.Mean({1}, false);
```
[the remark this answers is not in the document]
Comment (yxlao) [reply src/UnitTest/Core/Tensor.cpp:2308]: Tests added.
In general, the `keepdim=true` vs keepdim=false` is already tested in the `Sum()` op. `Mean() should have the same behavior as `Sum()`.

File: src/UnitTest/Python/test_core.py (hunk 1 above)
Comment (yxlao) [reply src/UnitTest/Python/test_core.py:548]: Done. Also added this to other python tests.

**Comment by germanros1987** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/UnitTest/Core/Tensor.cpp  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/intel-isl/Open3D/pullRequest?prid=5485035&bid=17973629)

File: src/UnitTest/Core/Tensor.cpp (hunk 2 above)
Comment (yxlao) [reply src/UnitTest/Core/Tensor.cpp:1338]: done, added in `test_core.py::test_reduction_special_shapes`

**Comment by yxlao** [comment]
Update: also added `pytest` to run on CI.

**Comment by germanros1987** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/UnitTest/Core/Tensor.cpp  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/intel-isl/Open3D/pullRequest?prid=5485035&bid=17973629)

**Comment by pvarvak** [comment]
Travis is failing on python 3.5 (because we don't support it, should we disable python 3.5 test?);
and also failing because ipywidgets is not installed (do we also need to update docs that ipywidgets needs to be installed to run python tests? both in UnitTest/Python/readme.md, and lets add it to docs/compilation.rst?)

**Comment by yxlao** [comment]
For the `ipywidgets` dependency, now i use `make install-pip-package` instead of `make install-python-package`. This should take care of installing all dependencies in `requirements.txt`. Thus no need to change the readme.

I believe German still wants the code to be compatible with Python 3.5. So I removed all f-strings in the repo. Now Python 3.5 should still work.

**Comment by germanros1987** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/UnitTest/Core/Tensor.cpp  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/intel-isl/Open3D/pullRequest?prid=5485035&bid=17973629)

**Comment by germanros1987** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/UnitTest/Core/Tensor.cpp  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/intel-isl/Open3D/pullRequest?prid=5485035&bid=17973629)

**Comment by germanros1987** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/UnitTest/Core/Tensor.cpp  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/intel-isl/Open3D/pullRequest?prid=5485035&bid=17973629)

**Comment by germanros1987** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/UnitTest/Core/Tensor.cpp  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/intel-isl/Open3D/pullRequest?prid=5485035&bid=17973629)

**Comment by germanros1987** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/UnitTest/Core/Tensor.cpp  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/intel-isl/Open3D/pullRequest?prid=5485035&bid=17973629)

**Comment by germanros1987** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/UnitTest/Core/Tensor.cpp  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/intel-isl/Open3D/pullRequest?prid=5485035&bid=17973629)

**Pull request merged**